File: typescript/packages/core/tests/verifier.test.ts
/**
 * Unit tests for {@link Verifier} (spore-core issue #44).
 *
 * Mirrors `rust/crates/spore-core/src/verifier.rs#tests` — same rules,
 * same verdicts, parallel structure.
 */

import { describe, expect, it } from "vitest";

import { SessionId, verifier } from "../src/index.js";
import type {
  CommandOutput,
  HaltReason,
  RunResult,
  SandboxProvider,
  SandboxViolation,
  ToolCall,
} from "../src/index.js";

const {
  CompositeVerifier,
  EvaluatorResponseVerifier,
  TestSuiteVerifier,
  COMPOSITE_REASON_CAP,
  DEFAULT_MAX_ITERATIONS,
  failed,
  passed,
} = verifier;

type Verifier = verifier.Verifier;
type VerifierInput = verifier.VerifierInput;
type VerifierVerdict = verifier.VerifierVerdict;

// ── helpers ──────────────────────────────────────────────────────────────────

function success(output: string): RunResult {
  return {
    kind: "success",
    output,
    session_id: SessionId.of("s"),
    usage: {
      input_tokens: 0,
      output_tokens: 0,
      cache_read_tokens: 0,
      cache_write_tokens: 0,
      cost_usd: 0,
    },
    turns: 1,
  };
}

function failure(
  reason: HaltReason = { kind: "strategy_not_yet_implemented", strategy: "x" },
): RunResult {
  return {
    kind: "failure",
    reason,
    session_id: SessionId.of("s"),
    usage: {
      input_tokens: 0,
      output_tokens: 0,
      cache_read_tokens: 0,
      cache_write_tokens: 0,
      cost_usd: 0,
    },
    turns: 0,
  };
}

function inputWith(build: RunResult, evalR: RunResult): VerifierInput {
  return {
    build_result: build,
    eval_result: evalR,
    workspace: "/tmp",
    iteration: 0,
  };
}

function makeRespVerifier(): verifier.EvaluatorResponseVerifier {
  return new EvaluatorResponseVerifier({
    pass_pattern: "(?i)\\bPASS\\b",
    fail_pattern: "(?i)\\bFAIL: .+",
  });
}

// ── EvaluatorResponseVerifier ────────────────────────────────────────────────

describe("EvaluatorResponseVerifier", () => {
  it("pass pattern matches → Passed", async () => {
    const v = makeRespVerifier();
    const i = inputWith(success("ok"), success("all checks PASS, ready to ship"));
    expect(await v.verify(i)).toEqual(passed());
  });

  it("fail pattern matches → Failed with reason", async () => {
    const v = makeRespVerifier();
    const i = inputWith(success("ok"), success("FAIL: missing edge case in handler.rs"));
    const verdict = await v.verify(i);
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason).toContain("missing edge case");
    }
  });

  it("neither pattern → default Failed mentioning output", async () => {
    const v = makeRespVerifier();
    const i = inputWith(success("ok"), success("indeterminate output"));
    const verdict = await v.verify(i);
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason).toContain("matched neither");
      expect(verdict.reason).toContain("indeterminate output");
    }
  });

  it("build failure propagates", async () => {
    const v = makeRespVerifier();
    const i = inputWith(failure(), success("PASS"));
    const verdict = await v.verify(i);
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason.startsWith("build run halted")).toBe(true);
    }
  });

  it("eval failure propagates", async () => {
    const v = makeRespVerifier();
    const i = inputWith(success("ok"), failure());
    const verdict = await v.verify(i);
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason.startsWith("evaluator run halted")).toBe(true);
    }
  });

  it("max_iterations defaults to 3 and is overridable", () => {
    const v = makeRespVerifier();
    expect(v.maxIterations()).toBe(DEFAULT_MAX_ITERATIONS);
    expect(v.maxIterations()).toBe(3);
    const v2 = new EvaluatorResponseVerifier({
      pass_pattern: "a",
      fail_pattern: "b",
      max_iterations: 10,
    });
    expect(v2.maxIterations()).toBe(10);
  });
});

// ── TestSuiteVerifier ────────────────────────────────────────────────────────

class StubSandbox implements SandboxProvider {
  constructor(private readonly out: CommandOutput) {}
  async validate(_call: ToolCall): Promise<SandboxViolation | null> {
    void _call;
    return null;
  }
  async executeCommand(): Promise<CommandOutput | SandboxViolation> {
    return this.out;
  }
}

function stubSandbox(exit: number, stderr: string): SandboxProvider {
  return new StubSandbox({
    stdout: "",
    stderr,
    exit_code: exit,
    timed_out: false,
    truncated: false,
  });
}

describe("TestSuiteVerifier", () => {
  it("exit 0 → Passed", async () => {
    const v = new TestSuiteVerifier({
      command: "pnpm test",
      working_dir: "/work",
      timeout_ms: 60_000,
      sandbox: stubSandbox(0, ""),
    });
    const i = inputWith(success("ok"), success(""));
    expect(await v.verify(i)).toEqual(passed());
  });

  it("non-zero exit → Failed includes stderr tail", async () => {
    const v = new TestSuiteVerifier({
      command: "pnpm test",
      working_dir: "/work",
      timeout_ms: 60_000,
      sandbox: stubSandbox(1, "test foo ... FAILED"),
    });
    const i = inputWith(success("ok"), success(""));
    const verdict = await v.verify(i);
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason).toContain("FAILED");
    }
  });

  it("build failure short-circuits", async () => {
    const v = new TestSuiteVerifier({
      command: "pnpm test",
      working_dir: "/work",
      timeout_ms: 60_000,
      sandbox: stubSandbox(0, ""),
    });
    const i = inputWith(failure(), success(""));
    const verdict = await v.verify(i);
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason.startsWith("build run halted")).toBe(true);
    }
  });

  it("empty command → Failed", async () => {
    const v = new TestSuiteVerifier({
      command: "",
      working_dir: "/work",
      timeout_ms: 60_000,
      sandbox: stubSandbox(0, ""),
    });
    const i = inputWith(success("ok"), success(""));
    const verdict = await v.verify(i);
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason).toContain("empty test command");
    }
  });
});

// ── CompositeVerifier ────────────────────────────────────────────────────────

class FixedVerifier implements Verifier {
  constructor(private readonly verdict: VerifierVerdict) {}
  async verify(): Promise<VerifierVerdict> {
    return this.verdict;
  }
  maxIterations(): number {
    return DEFAULT_MAX_ITERATIONS;
  }
}

const passV = (): Verifier => new FixedVerifier(passed());
const failV = (reason: string): Verifier => new FixedVerifier(failed(reason));

describe("CompositeVerifier", () => {
  it("all pass → Passed", async () => {
    const c = new CompositeVerifier({ verifiers: [passV(), passV(), passV()] });
    expect(await c.verify(inputWith(success("ok"), success("ok")))).toEqual(passed());
  });

  it("one fail → Failed with that reason and tagged index", async () => {
    const c = new CompositeVerifier({ verifiers: [passV(), failV("oops"), passV()] });
    const verdict = await c.verify(inputWith(success("ok"), success("ok")));
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason).toContain("oops");
      expect(verdict.reason).toContain("[verifier 1]");
    }
  });

  it("many fails concatenated, passing children not mentioned", async () => {
    const c = new CompositeVerifier({
      verifiers: [failV("first"), passV(), failV("second"), failV("third")],
    });
    const verdict = await c.verify(inputWith(success("ok"), success("ok")));
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason).toContain("first");
      expect(verdict.reason).toContain("second");
      expect(verdict.reason).toContain("third");
      expect(verdict.reason).not.toContain("[verifier 1]");
    }
  });

  it("truncates at 2000 chars", async () => {
    const long = "x".repeat(5000);
    const c = new CompositeVerifier({ verifiers: [failV(long)] });
    const verdict = await c.verify(inputWith(success("ok"), success("ok")));
    expect(verdict.kind).toBe("failed");
    if (verdict.kind === "failed") {
      expect(verdict.reason.length).toBeLessThanOrEqual(
        COMPOSITE_REASON_CAP + "... [truncated]".length,
      );
      expect(verdict.reason.endsWith("... [truncated]")).toBe(true);
    }
  });

  it("max_iterations honored", () => {
    const c = new CompositeVerifier({ verifiers: [passV()], max_iterations: 7 });
    expect(c.maxIterations()).toBe(7);
  });
});

File: go/spore-core/verifier/verifier_test.go
package verifier

import (
	"context"
	"strings"
	"testing"
	"time"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

func successResult(output string) sporecore.RunResult {
	return sporecore.RunResult{
		Kind:      sporecore.RunSuccess,
		Output:    output,
		SessionID: sporecore.SessionID("s"),
		Turns:     1,
	}
}

func failureResult() sporecore.RunResult {
	return sporecore.RunResult{
		Kind: sporecore.RunFailure,
		Reason: sporecore.HaltReason{
			Kind:     sporecore.HaltStrategyNotYetImplemented,
			Strategy: "x",
		},
		SessionID: sporecore.SessionID("s"),
	}
}

func waitingForHumanResult() sporecore.RunResult {
	return sporecore.RunResult{Kind: sporecore.RunWaitingForHuman}
}

func inputWith(build, eval sporecore.RunResult) *VerifierInput {
	return &VerifierInput{
		BuildResult: build,
		EvalResult:  eval,
		Workspace:   "/tmp",
		Iteration:   0,
	}
}

func makeRespVerifier(t *testing.T) *EvaluatorResponseVerifier {
	t.Helper()
	v, err := NewEvaluatorResponseVerifier(`(?i)\bPASS\b`, `(?i)\bFAIL: .+`, 3)
	if err != nil {
		t.Fatalf("compile: %v", err)
	}
	return v
}

// ── EvaluatorResponseVerifier ────────────────────────────────────────────────

func TestResponseVerifierPassPatternMatches(t *testing.T) {
	v := makeRespVerifier(t)
	in := inputWith(successResult("ok"), successResult("all checks PASS, ready to ship"))
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictPassed {
		t.Fatalf("expected Passed, got %+v", got)
	}
}

func TestResponseVerifierFailPatternMatchesWithReason(t *testing.T) {
	v := makeRespVerifier(t)
	in := inputWith(successResult("ok"), successResult("FAIL: missing edge case in handler.rs"))
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if !strings.Contains(got.Reason, "missing edge case") {
		t.Fatalf("expected reason to contain 'missing edge case', got %q", got.Reason)
	}
}

func TestResponseVerifierNeitherPatternDefaultFails(t *testing.T) {
	v := makeRespVerifier(t)
	in := inputWith(successResult("ok"), successResult("indeterminate output"))
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if !strings.Contains(got.Reason, "matched neither") {
		t.Fatalf("expected 'matched neither' in reason, got %q", got.Reason)
	}
	if !strings.Contains(got.Reason, "indeterminate output") {
		t.Fatalf("expected eval output in reason, got %q", got.Reason)
	}
}

func TestResponseVerifierBuildFailurePropagates(t *testing.T) {
	v := makeRespVerifier(t)
	in := inputWith(failureResult(), successResult("PASS"))
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if !strings.HasPrefix(got.Reason, "build run halted") {
		t.Fatalf("expected 'build run halted' prefix, got %q", got.Reason)
	}
}

func TestResponseVerifierEvalFailurePropagates(t *testing.T) {
	v := makeRespVerifier(t)
	in := inputWith(successResult("ok"), failureResult())
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if !strings.HasPrefix(got.Reason, "evaluator run halted") {
		t.Fatalf("expected 'evaluator run halted' prefix, got %q", got.Reason)
	}
}

func TestResponseVerifierWaitingForHumanIsMisconfiguration(t *testing.T) {
	v := makeRespVerifier(t)
	in := inputWith(successResult("ok"), waitingForHumanResult())
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if !strings.Contains(got.Reason, "WaitingForHuman") {
		t.Fatalf("expected WaitingForHuman in reason, got %q", got.Reason)
	}
	if !strings.Contains(got.Reason, "misconfiguration") {
		t.Fatalf("expected misconfiguration mention, got %q", got.Reason)
	}
}

func TestResponseVerifierMaxIterations(t *testing.T) {
	v := makeRespVerifier(t)
	if v.MaxIterations() != 3 {
		t.Fatalf("expected 3, got %d", v.MaxIterations())
	}
	v2, err := NewEvaluatorResponseVerifier("a", "b", 10)
	if err != nil {
		t.Fatal(err)
	}
	if v2.MaxIterations() != 10 {
		t.Fatalf("expected 10, got %d", v2.MaxIterations())
	}
}

func TestResponseVerifierInvalidRegexErrors(t *testing.T) {
	if _, err := NewEvaluatorResponseVerifier("(unclosed", "b", 3); err == nil {
		t.Fatalf("expected error for invalid pass_pattern")
	}
	if _, err := NewEvaluatorResponseVerifier("a", "(unclosed", 3); err == nil {
		t.Fatalf("expected error for invalid fail_pattern")
	}
}

// ── TestSuiteVerifier ────────────────────────────────────────────────────────

type stubSandbox struct {
	out  sporecore.CommandOutput
	root string
}

func (s *stubSandbox) Validate(context.Context, sporecore.ToolCall) *sporecore.SandboxViolation {
	return nil
}
func (s *stubSandbox) ExecuteCommand(
	ctx context.Context,
	command string,
	args []string,
	workingDir string,
	timeout time.Duration,
) (sporecore.CommandOutput, *sporecore.SandboxViolation) {
	return s.out, nil
}
func (s *stubSandbox) HandleLargeOutput(
	ctx context.Context, content, callID string, head, tail uint32,
) sporecore.TruncatedOutput {
	return sporecore.TruncatedOutput{Content: content}
}
func (s *stubSandbox) ResolvePath(
	ctx context.Context, p string, op sporecore.Operation,
) (string, *sporecore.SandboxViolation) {
	return p, nil
}
func (s *stubSandbox) IsolationMode() sporecore.IsolationMode {
	return sporecore.IsolationNone{}
}
func (s *stubSandbox) WorkspaceRoot() string { return s.root }

func newStubSandbox(exit int, stderr string) sporecore.SandboxProvider {
	return &stubSandbox{
		out: sporecore.CommandOutput{
			Stdout:   "",
			Stderr:   stderr,
			ExitCode: exit,
			TimedOut: false,
		},
		root: "/",
	}
}

func TestSuiteVerifierPass(t *testing.T) {
	v := NewTestSuiteVerifier("cargo test", "/work", 60*time.Second, newStubSandbox(0, ""), 3)
	in := inputWith(successResult("ok"), successResult(""))
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictPassed {
		t.Fatalf("expected Passed, got %+v", got)
	}
}

func TestSuiteVerifierFailIncludesStderr(t *testing.T) {
	v := NewTestSuiteVerifier("cargo test", "/work", 60*time.Second, newStubSandbox(1, "test foo ... FAILED"), 3)
	in := inputWith(successResult("ok"), successResult(""))
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if !strings.Contains(got.Reason, "FAILED") {
		t.Fatalf("expected FAILED in reason, got %q", got.Reason)
	}
}

func TestSuiteVerifierBuildFailureShortCircuits(t *testing.T) {
	v := NewTestSuiteVerifier("cargo test", "/work", 60*time.Second, newStubSandbox(0, ""), 3)
	in := inputWith(failureResult(), successResult(""))
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if !strings.HasPrefix(got.Reason, "build run halted") {
		t.Fatalf("expected 'build run halted' prefix, got %q", got.Reason)
	}
}

func TestSuiteVerifierEmptyCommandFails(t *testing.T) {
	v := NewTestSuiteVerifier("", "/work", 60*time.Second, newStubSandbox(0, ""), 3)
	in := inputWith(successResult("ok"), successResult(""))
	got := v.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if !strings.Contains(got.Reason, "empty test command") {
		t.Fatalf("expected 'empty test command', got %q", got.Reason)
	}
}

func TestSuiteVerifierMaxIterations(t *testing.T) {
	v := NewTestSuiteVerifier("x", "/work", time.Second, newStubSandbox(0, ""), 7)
	if v.MaxIterations() != 7 {
		t.Fatalf("expected 7, got %d", v.MaxIterations())
	}
}

// ── CompositeVerifier ────────────────────────────────────────────────────────

type fixedVerifier struct {
	v VerifierVerdict
}

func (f *fixedVerifier) Verify(context.Context, *VerifierInput) VerifierVerdict { return f.v }
func (f *fixedVerifier) MaxIterations() uint32                                  { return 3 }

func passV() Verifier              { return &fixedVerifier{v: Passed()} }
func failV(reason string) Verifier { return &fixedVerifier{v: Failed(reason)} }

func TestCompositeAllPassReturnsPassed(t *testing.T) {
	c := NewCompositeVerifier([]Verifier{passV(), passV(), passV()}, 3)
	in := inputWith(successResult("ok"), successResult("ok"))
	got := c.Verify(context.Background(), in)
	if got.Kind != VerdictPassed {
		t.Fatalf("expected Passed, got %+v", got)
	}
}

func TestCompositeOneFailReturnsThatReason(t *testing.T) {
	c := NewCompositeVerifier([]Verifier{passV(), failV("oops"), passV()}, 3)
	in := inputWith(successResult("ok"), successResult("ok"))
	got := c.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if !strings.Contains(got.Reason, "oops") {
		t.Fatalf("expected 'oops' in reason, got %q", got.Reason)
	}
	if !strings.Contains(got.Reason, "[verifier 1]") {
		t.Fatalf("expected '[verifier 1]' in reason, got %q", got.Reason)
	}
}

func TestCompositeManyFailsConcatenated(t *testing.T) {
	c := NewCompositeVerifier(
		[]Verifier{failV("first"), passV(), failV("second"), failV("third")}, 3,
	)
	in := inputWith(successResult("ok"), successResult("ok"))
	got := c.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	for _, sub := range []string{"first", "second", "third"} {
		if !strings.Contains(got.Reason, sub) {
			t.Fatalf("expected %q in reason, got %q", sub, got.Reason)
		}
	}
	if strings.Contains(got.Reason, "[verifier 1]") {
		t.Fatalf("pass-verifier should not appear in reason: %q", got.Reason)
	}
}

func TestCompositeTruncatesAt2000Chars(t *testing.T) {
	long := strings.Repeat("x", 5000)
	c := NewCompositeVerifier([]Verifier{failV(long)}, 3)
	in := inputWith(successResult("ok"), successResult("ok"))
	got := c.Verify(context.Background(), in)
	if got.Kind != VerdictFailed {
		t.Fatalf("expected Failed, got %+v", got)
	}
	if len(got.Reason) > compositeReasonCap+len("... [truncated]") {
		t.Fatalf("reason exceeded cap: len=%d", len(got.Reason))
	}
	if !strings.HasSuffix(got.Reason, "... [truncated]") {
		t.Fatalf("expected truncation suffix, got %q", got.Reason[len(got.Reason)-30:])
	}
}

func TestCompositeMaxIterations(t *testing.T) {
	c := NewCompositeVerifier(nil, 9)
	if c.MaxIterations() != 9 {
		t.Fatalf("expected 9, got %d", c.MaxIterations())
	}
}

// Compile-time guarantee that the Verifier interface is satisfied by an
// interface-typed value.
func TestVerifierInterfaceUsable(t *testing.T) {
	v, err := NewEvaluatorResponseVerifier("PASS", "FAIL", 3)
	if err != nil {
		t.Fatal(err)
	}
	var iface Verifier = v
	_ = iface.MaxIterations()
}
